Chore: NestJS 스켈레톤 프로젝트 초기 설정

Zamoca42 opened the pull request:
## PR 체크리스트
아래 항목을 확인해 주세요:

- [x] 커밋 메시지가 우리의 가이드라인을 따르고 있는지 확인하세요
- [ ] 변경 사항에 대한 테스트가 추가되었는지 확인하세요 (버그 수정 / 기능 추가)
- [ ] 문서가 추가되거나 업데이트되었는지 확인하세요 (버그 수정 / 기능 추가)

## PR 유형
이 PR은 어떤 종류의 변경을 가져오나요?

<!-- 이 PR에 해당하는 것을 "x"로 표시하세요. -->

- [ ] 버그 수정
- [ ] 새로운 기능 추가
- [ ] 코드 스타일 업데이트 (서식, 로컬 변수)
- [ ] 리팩터링 (기능 변경 없음, API 변경 없음)
- [x] 빌드 관련 변경
- [x] CI 관련 변경
- [x] 문서 내용 변경
- [x] 애플리케이션 / 인프라 변경
- [ ] 기타... 설명:

## 현재 동작은 무엇인가요?
<!-- 수정 중인 현재 동작을 설명하거나 관련된 이슈에 링크하세요. -->

이슈 번호: #1 

## 새로운 동작은 무엇인가요?

## 이 PR은 호환성 변경을 도입하나요?

- [ ] 예
- [x] 아니요

<!-- 이 PR에 호환성 변경이 포함되어 있다면, 기존 응용 프로그램에 대한 영향과 마이그레이션 경로를 아래에 설명해 주세요. -->

## 기타 정보

Zamoca42 committed:
Chore: NestJS 프로젝트 시작

- `nest new ./` 명령어를 사용하여 NestJS 프로젝트 초기 설정을 적용

Related to #1

Zamoca42 committed:
Docs: 템플릿 추가

- 일관성 있는 협업을 위해 PR/버그/이슈/커밋 템플릿 추가
Related to #1

Zamoca42 committed:
Chore: 도커, 도커 컴포즈 설정 추가

- 개발환경을 일관되게 설정할 수 있도록 도커 컨테이너 설정 적용

Related to #1

Zamoca42 committed:
Chore: 패키지 적용

- dto에서 검증, 변환 작업을 위해 class-vt,tf 패키지 적용
- api 문서화 작업을 위해 swagger 패키지 적용

Related to #1

Zamoca42 committed:
Ci: GitHub-Actions CI 설정 적용

- PR시 포맷팅, 테스트, 린트가 통과했는지 확인하기 위해 CI 설정 적용

Related to #1

Zamoca42 committed:
Chore: 환경변수 컨트롤 및 검증 설정 적용 #1

- 일관성 있는 개발환경을 위해서 환경변수 컨트롤하고 검증하는 미들웨어를 설정

Related to #1

Zamoca42 committed:
Feat: global exception filter 적용 #1

- 서비스 로직에서 간단한 예외처리를 위해 전역 예외 필터 설정 적용

Related to #1

Zamoca42 committed:
Chore: 필요 없는 환경변수 검증 제거 #1

Related to #1

Zamoca42 committed:
Chore: swagger 설정 적용 #1

- api 문서화를 위해서 swagger 설정 적용

Related to #1

Zamoca42 committed:
Docs: 개발 문서 변경

- 스켈레톤 프로젝트를 사용해보기 위한 실행 방법 적용 (임시)

Related to #1

myeongseoklee commented on .gitignore:
```diff
@@ -0,0 +1,35 @@
+!.vscode/settings.json
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
```
env 파일도 추가되어야 할 것 같습니다!

myeongseoklee commented on package.json:
```diff
@@ -0,0 +1,69 @@
+    "ts-node": "^10.9.1",
+    "tsconfig-paths": "^4.2.0",
+    "typescript": "^5.1.3"
+  },
```
```suggestion
{"packageManager": "yarn@1.22.19",
	"engines": {
		"node": "18.16.1"
	},
```

제가 전에 개인프로젝트 할 때 노드버전과 패키지매니저 버전을 고정하려고 사용했었는데요.
버전 안정성을 위해 노드 버전을 명확히 지정하거나, 몇 버전 이상의 버전을 사용해야 한다는 등 규칙을 적을 수 있습니다.
docker 이미지가 돌아가는 환경의 노드버전과 일치시키면 어떨까 싶습니다.(가능하다면 패키지 매니저버전까지)

저같은 경우는 node 18.16 버전을 사용중인데, docker에 설정하신 노드버전을 보면 20이더라구요.
다른 분들도 노드버전이 다르시지 않을까 싶어, 명시해두면 좋을 것 같아 제안드립니다.

참고로 로컬에서 node 버전은 매번 바꿔주기 귀찮을 수 있으니 프로젝트의 루트에 .nvmrc 파일 추가하고 몇 가지 쉘 스크립트를 작성하면 패키지별로 지정한 노드버전이 다르더라도 해당 버전으로 작업이 됩니다.
(참고: https://hyeok999.github.io/2020/06/02/NVM/#a5)

Zamoca42 replied:
이 부분에서는 저와 의도는 같은데 방법은 다른거 같아서 코멘트 남깁니다. 

제가 docker compose를 사용한 부분은 로컬 환경에서 노드 버전이 모두 다르기 때문에 `docker compose up`을 사용하게 되면 node 20 버전으로 도커 환경 실행되니까 모두 같은 노드 버전의 환경에서 작업해볼 수 있기 때문에 도커를 사용했습니다.

로컬에서까지 .nvmrc로 노드 버전을 맞춰서 작업하게되면 docker가 의미가 없게 되는거 같습니다.

myeongseoklee replied:
아 네넵 그부분은 이해했습니다.

그럼 도커 컨테이너 위에서 의존성 패키지를 npm install 하게되면 도커 컨테이너의 node version에 맞게 의존성 버전이 제대로 설치된다는 말씀이시죠?

저는 그 부분까지는 몰랐고, 도커를 사용하지 않으면 npm install은 로컬의 node version을 따라가니까 일치해야하지 않나 생각했었어요😅

그렇다면, "꼭" 도커 컨테이너를 띄우고 작업해야겠군요!

myeongseoklee commented on docker/local.Dockerfile:
```diff
@@ -0,0 +1,20 @@
+COPY package*.json ./
+
+# Install application dependencies
+RUN npm install
```
```suggestion
Run npm ci
```
"npm i"는 실행 시, package.json에 ^ 또는 ~를 사용하는 경우 지정된 버전과 동일한 버전을 설치하지 못할 수 있다고 합니다.
따라서, 버전이 달라지면 package-lock.json이 업데이트 되므로 새로운 의존성을 설치할 때가 아니라
1. CI/CD 과정에서 전체 의존성을 설치 하거나
2. 프로젝트 레포지토리를 처음 clone 받아 의존성을 설치할 때는
"npm ci"를 사용하는 것이 좋다고 합니다.
(참고: https://doqtqu.tistory.com/353)

Zamoca42 replied:
ci.yml과 dockerfile에서 npm install -> ci로 변경했습니다

Zamoca42 replied:
네 이부분은 npm install -> ci로 변경해보겠습니다

myeongseoklee commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,38 @@
+on:
+  pull_request:
+    branches:
+      - 'develop'
```
저희 develop 브랜치가 없어서 삭제해도 될 것 같습니다.

Zamoca42 replied:
넵 TIL 시간에 논의했던 부분이라 제외하겠습니다.

Zamoca42 replied:
ci에서 develop 브랜치를 제외하고 feature 브랜치만 명시했습니다

myeongseoklee commented on package.json:
```diff
@@ -8,10 +8,10 @@
-    "start": "nest start",
-    "start:dev": "nest start --watch",
+    "start": "cross-env NODE_ENV=local nest start",
+    "start:dev": "cross-env NODE_ENV=development nest start",
```
저희 개발서버가 따로 없어서, dev env 주입해서 서버를 킬 일이 있을까 싶어요!

Zamoca42 replied:
요것도 예시상황으로 넣어봤는데 환경부분 검증에서 제외할 필요가 있겠네요

Zamoca42 replied:
dev env를 제외하고 env.example로 환경 변수 예시를 추가했습니다.

myeongseoklee commented on src/common/env.validation.ts:
```diff
@@ -0,0 +1,35 @@
+
+  @IsString()
+  DYNAMODB_TABLE_PREFIX: string;
+}
```
Q. env에서 사용할 필드를 class의 멤버변수로 추가해주어야 하나요?

DYNAMODB_TABLE_PREFIX 멤버변수는 필요한 값이 아니면 삭제해야 할 것 같습니다.

myeongseoklee commented on src/common/filter/http-exception.filter.ts:
```diff
@@ -0,0 +1,41 @@
+      }`,
+    );
+
+    if (exception instanceof BadRequestException) {
+      response.status(exception.getStatus()).json(exception);
+      return;
+    }
```
여기서 BadRequestException인 경우만 따로 처리하신 이유가 있을까요?

Zamoca42 replied:
저번 프로젝트에서 그대로 exception filter를 가져와봤는데, 저번 프로젝트에서는 프론트에서 400error를 직접 보내는일이 많아서 저렇게 처리해줬었던걸로 기억합니다. 일반화할 때는 필요없는 부분인거 같아서 이번에는 제외하겠습니다.

myeongseoklee replied:
```suggestion
switch (exception.constructor) {
      case HttpException: // for HttpException
        status = (exception as HttpException).getStatus();
        break;

      case QueryFailedError: // for TypeOrm error
        status = HttpStatus.UNPROCESSABLE_ENTITY;
        message = (exception as QueryFailedError).message;
        code = (exception as any).code;
        break;

      case BadRequestException:
        status = (exception as BadRequestException).getStatus();
        break;

      case UnauthorizedException: // for Auth guard error
        status = (exception as UnauthorizedException).getStatus();
        break;

      case ForbiddenException:
        status = (exception as ForbiddenException).getStatus();
        break;

      case NotFoundException:
        status = (exception as NotFoundException).getStatus();
        break;

      case ConflictException:
        status = (exception as ConflictException).getStatus();
        break;

      default: // default
        status = HttpStatus.INTERNAL_SERVER_ERROR;
    }
```

ㅋㅋ 저도 실은 전 프로젝트에서 exception filter에 에러별로 status 할당하는 메소드를 사용하긴 했었는데요, 이번에 다시 보니 BadRequestException 등 여러 exception의 부모 클래스가 HttpException이더라구요. 그래서 사실상 제가 만들어 썼던 메소드가 큰 의미가 없었다는 것을 깨닫게 되어서, 혹시 연규님은 다른 의도가 있나 궁금해서 여쭤봤습니다!

Zamoca42 replied:
exception filter에서 BadRequest(400)만 따로 처리하는 경우를 제외했습니다

myeongseoklee commented on src/main.ts:
```diff
@@ -1,8 +1,28 @@
 async function bootstrap() {
   const app = await NestFactory.create(AppModule);
+  const allowedOrigins = process.env.ALLOWED_ORIGINS.split(',');
+
```
```suggestion
  app.useGlobalInterceptors(new ClassSerializerInterceptor(app.get(Reflector)));
```
위 코드를 혹시 추가해주실 수 있을까요? 응답객체 직렬화에 사용합니다.
(참고: https://jojoldu.tistory.com/610)

Zamoca42 replied:
네 참고링크가 많은 도움되었습니다! 이 부분도 추가해보겠습니다

Zamoca42 replied:
main.ts에서 글로벌 인터셉터를 추가했습니다

myeongseoklee reviewed (changes requested):
몇 가지 제안과 수정 요청을 남겼습니다. 오늘 고생 많으셨습니다!

Zamoca42 committed:
Chore: 개발 환경을 제외하고 예시 환경 추가

- 개발 서버가 필요 없기 때문에 development 환경 제외
- env.example을 추가해서 환경 변수 예시에 대해 명시

Related to #1

Zamoca42 committed:
Chore: 패키지 설치 명령 수정

- 같은 패키지 버전을 맞추기 위해 npm install -> ci로 변경

Related to #1

Zamoca42 committed:
Refactor: 예외 필터 수정

- 예외 필터 일반화를 위해 BadRequest(400) 에러만 따로 처리하던 경우 제외

Related to #1

Zamoca42 committed:
Feat: 응답객체 직렬화 추가

- 응답객체 직렬화를 위한 글로벌 인터셉터 추가

- Related to #1

Zamoca42 committed:
Docs: 로컬 환경 실행 명령 제외

- 로컬 환경 실행을 제외하고 노드 버전을 통일하기 위해 도커 실행만 명시

Related to #1

hojoonSong commented:
node환경이 오랜만이고, docker compose와 각종 의존성들을 미리 합의하는 과정 잘 보고 배웠습니다!

refigo reviewed (approved):
초기 설정 확인했습니다!
